### PR TITLE
ISPN-4546, ISPN-5274 - Handle more corner cases when retrying tx commands

### PR DESCRIPTION
**Pull request by danberindei**
https://issues.jboss.org/browse/ISPN-4546
https://issues.jboss.org/browse/ISPN-5274
https://issues.jboss.org/browse/ISPN-5399

* Add more tests
* Handle LockControlCommands
* Avoid retrying CommitCommands after the remote transactions were removed

### CODE REVIEW

File: core/src/main/java/org/infinispan/commands/tx/AbstractTransactionBoundaryCommand.java [hunk 1]
```diff
@@ -5,6 +5,7 @@
 import org.infinispan.context.impl.RemoteTxInvocationContext;
 import org.infinispan.interceptors.InterceptorChain;
 import org.infinispan.lifecycle.ComponentStatus;
+import org.infinispan.remoting.responses.UnsureResponse;
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/commands/tx/AbstractTransactionBoundaryCommand.java]: not used

File: core/src/main/java/org/infinispan/interceptors/distribution/TxDistributionInterceptor.java [hunk 2]
```diff
@@ -207,13 +214,13 @@ protected void prepareOnAffectedNodes(TxInvocationContext<?> ctx, PrepareCommand
       try {
          // this method will return immediately if we're the only member (because exclude_self=true)
          RpcOptions rpcOptions;
-         if (sync && command.isOnePhaseCommit()) {
+         if (sync) {
             rpcOptions = rpcManager.getRpcOptionsBuilder(ResponseMode.SYNCHRONOUS_IGNORE_LEAVERS, DeliverOrder.NONE).build();
          } else {
             rpcOptions = rpcManager.getDefaultRpcOptions(sync);
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/interceptors/distribution/TxDistributionInterceptor.java]: can replace ```sync``` by ```false```

File: core/src/main/java/org/infinispan/interceptors/distribution/VersionedDistributionInterceptor.java [hunk 3]
```diff
@@ -34,7 +34,7 @@ protected void prepareOnAffectedNodes(TxInvocationContext<?> ctx, PrepareCommand
       // Perform the RPC
       try {
          Map<Address, Response> resps = rpcManager.invokeRemotely(recipients, command, rpcManager.getDefaultRpcOptions(true, DeliverOrder.NONE));
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/interceptors/distribution/VersionedDistributionInterceptor.java]: don't you want to use  ```SYNCHRONOUS_IGNORE_LEAVERS```?

File: core/src/main/java/org/infinispan/transaction/impl/TransactionTable.java [hunk 4]
```diff
@@ -667,7 +689,7 @@ public void cleanupCompletedTransactions() {
             int removedEntries = 0;
 
             // Collect the leavers. They will be removed at the end.
-            Set<Address> leavers = new HashSet<>();
+            Set<Address> leavers = new HashSet<Address>();
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/transaction/impl/TransactionTable.java]: we are in java 7 :)

File: core/src/test/java/org/infinispan/partitionhandling/NonTxPartitionStressTest.java [hunk 5]
```diff
@@ -27,17 +31,20 @@
 import static org.testng.AssertJUnit.assertFalse;
 
 @Test(groups = "stress", testName = "partitionhandling.org.infinispan.partitionhandling.PartitionStressTest")
```
Comment (pruivo) [inline core/src/test/java/org/infinispan/partitionhandling/NonTxPartitionStressTest.java]: old test name

File: core/src/test/java/org/infinispan/partitionhandling/OptimisticPartitionStressTest.java [hunk 6]
```diff
@@ -0,0 +1,14 @@
+package org.infinispan.partitionhandling;
+
+import org.infinispan.transaction.LockingMode;
+import org.infinispan.transaction.TransactionMode;
+import org.testng.annotations.Test;
+
+@Test(groups = "stress", testName = "partitionhandling.org.infinispan.partitionhandling.OptimisticPartitionStressTest")
```
Comment (pruivo) [inline core/src/test/java/org/infinispan/partitionhandling/OptimisticPartitionStressTest.java]: wrong test name

File: core/src/test/java/org/infinispan/partitionhandling/PessimisticPartitionStressTest.java [hunk 7]
```diff
@@ -0,0 +1,17 @@
+package org.infinispan.partitionhandling;
+
+import org.infinispan.transaction.LockingMode;
+import org.infinispan.transaction.TransactionMode;
+import org.testng.annotations.Test;
+
+import static org.testng.AssertJUnit.assertEquals;
+import static org.testng.AssertJUnit.assertFalse;
+
+@Test(groups = "stress", testName = "partitionhandling.org.infinispan.partitionhandling.PessimisticPartitionStressTest")
```
Comment (pruivo) [inline core/src/test/java/org/infinispan/partitionhandling/PessimisticPartitionStressTest.java]: wrong test name

File: core/src/test/java/org/infinispan/partitionhandling/PessimisticPartitionStressTest.java [hunk 8]
```diff
@@ -0,0 +1,17 @@
+package org.infinispan.partitionhandling;
+
+import org.infinispan.transaction.LockingMode;
+import org.infinispan.transaction.TransactionMode;
+import org.testng.annotations.Test;
+
+import static org.testng.AssertJUnit.assertEquals;
```
Comment (pruivo) [inline core/src/test/java/org/infinispan/partitionhandling/PessimisticPartitionStressTest.java]: unused imports

File: core/src/main/java/org/infinispan/interceptors/distribution/VersionedDistributionInterceptor.java (hunk 3 above)
Comment (danberindei) [inline core/src/main/java/org/infinispan/interceptors/distribution/VersionedDistributionInterceptor.java]: No. If the prepare is synchronous then I want to throw an OutdatedTopologyException when a node is suspected or a cache isn't running (`checkTxCommandResponses(resps, command)` below), and to retry the command later.

**Comment by danberindei** [comment]
Updated

File: core/src/main/java/org/infinispan/interceptors/distribution/VersionedDistributionInterceptor.java (hunk 3 above)
Comment (pruivo) [inline core/src/main/java/org/infinispan/interceptors/distribution/VersionedDistributionInterceptor.java]: when a node is suspected, sync will throw a ```SuspectedException``` :)

Comment (danberindei) [inline core/src/main/java/org/infinispan/interceptors/distribution/VersionedDistributionInterceptor.java]: Fixed :)

**Comment by pruivo** [comment]
testing and pushing....

**Comment by pruivo** [comment]
integrated! thanks @danberindei!

**Pull request closed without merging**